<a href="https://colab.research.google.com/github/myrellamedvet/CursoRVigSaude/blob/atualizacaocursoRparavigilancia/cursorelementarbmclimafundamentos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


---

**<h1>CURSO BÁSICO DE LINGUAGEM R</h1>**

---

# Fundamentos de programação


Neste capítulo do curso, vamos abordar os fundamentos da programação aplicados ao R.

Programar um computador significa dizer a ele por meio de um conjunto de textos especiais (códigos) o que desejamos que seja feito. Ao escrever um código (script) precisamos ter em mente que o computador só fará aquilo que mandamos. E, como uma máquina, suas capacidades de compreensão são bem limitadas. Por isso, usamos algumas técnicas para que possamos nos fazer "entender" pelo computador.

A melhor parte é que quando aprendemos a "dizer" para o computador o que queremos, somos capazes de automatizar tarefas repetitivas, processar grandes quantidades de dados e realizar cálculos longos.

Mantenha em mente que o que queremos é entrar com os dados, que o computador processe conforme solicitamos e nos devolva a saída de dados ou informações de acordo com o desejado.

*ENTRADA -> PROCESSAMENTO -> SAÍDA*

No R utilizamos essas mesmas técnicas. Nas próximas sessões daremos as orientações e exemplos aplicados no próprio R para que você possa entender e praticar.


## Variáveis e constantes:

Para que um programa de computador funcione, é precisa entrar com os dados que desejamos. Isso pode ser feito duas maneiras:

1. Inserir manualmente cada campo.
2. Carregar um conjunto de dados.

Um exemplo de inserção manual são as calculadoras, onde colocamos um número, depois inseridos a operação que queremos e depois outro número e assim por diante.

Um exemplo de carga de dados é seu extrato bancário. Já notou que o extrato é um relatório de todas as transações registradas em certo período de tempo?

Da mesma maneira, os dados coletados durante atendimentos em uma UBS, notificados e depois disponibilizados via DataSUS também são exemplos de registros que podem se tornar relatórios.

É aí que entra a segunda parte, o processamento. Quando você coloca o seu login e senha no aplicativo do banco e solicita o extrato, o computador processa a sua solicitação e imprime os valores na tela ou gera um PDF. Da mesma maneira, com o R podemos construir scripts que nos respondam à perguntas que possuímos sobre os registros disponíveis nas bases de dados de clima e de saúde.

Então, é possível solicitar ao computador que realize as operações necessárias com os dados carregados? Sim. Cada vez que executamos (rodamos) um script, acontece um processamento.

O relatório é a saída do processamento. O seu extrato bancário possui operações consideradas simples, pois trata-se de uma consulta com adição e subtração das operações. Compreender as relações entre a temperatura do ar e morbimortalidade nos exigirá um pouco mais de esforço, mas o fundamento é sempre o mesmo.




Rode (execute) o script abaixo e veja a diferença no valor de x, a cada vez que você atribui um novo valor. (Você pode usar o RStudio ou o Google Colab.)

In [ ]:
x<-5
print(x)

[1] 5


In [ ]:
x<-10
print(x)

[1] 10


In [ ]:
x<-4500
print(x)

[1] 4500


Notou que cada vez que você muda o valor à sua direita, o valor em tela também muda? Por isso chamamos o x, ou qualquer outra letra ou nome à sua esquerda de variável.

Seja qual for a operação que você está fazendo no computador (entrada, processamento ou saída), será necessário "guardar" temporariamente os valores utilizados no cálculo ou consulta. Para isso dizemos ao computador que queremos "declarar uma variável".

Uma variável é uma unidade representada por letra ou números que pode ter seu valor modificado ao longo do processamento. Para isso usamos algo que você certamente conhece, equações.

Vamos a um exemplo:

*x = 3*

Se você ler a expressão acima em voz alta, deverá dizer que *xis é igual a três*. Não é mesmo?

No R, fazemos algo muito parecido, apenas precisamos deixar diferente a igualdade porque na verdade estamos atribuindo um valor para x que poderá ser modificado. Por isso no R utilizamos o símbolo *<-* para dizer que naquele momento *xis é igual a três*. Segue o exemplo:

*x<-3*

Pronto, você disse ao R que *a xis está atribuido o número inteiro três*, mas que isso pode mudar. Foi exatamente isso que acabamos de fazer nos exemplos acima. Utilizamos números, mas podemos atribuir qualquer caractere ou conjunto de dados a uma variável. Veremos isso com detalhes mais à frente.

E quando precisamos que um valor atribuído não seja modificado durante o processamento? Utilizamos as constantes. Como o próprio nome diz, ao contrário da variável ela não poderá ter seu valor atualizado.

As constantes são importantes pela clareza, facilidade de manutenção, prevenção de erros, segurança e melhoria no desempenho dos nossos scripts.

No R é comum utilizarmos funções para declarar constantes como o PI e E (constante de euler) ou valores vazios e nulos. No final dessa seção você vai encontrar uma lista com os tipos de contantes mais utilizadas pela linguagem.

Esses valores, por convenção, não serão modificados então os declaramos da seguinte forma:

```
PI <- 3.14159
E <- 2.71828
```

Por boas práticas, as variáveis sempre são declaradas com letras minúsculas e as constantes com letras maiúsculas.

Rode os exemplos abaixo e veja como se comportam as constantes.

In [ ]:
PI <- 3.14159
E <- 2.71828
print(PI)
print(E)

[1] 3.14159
[1] 2.71828


## Operadores

Agora que já sabemos o que são variávis e constantes, assim como declará-las utilizando o R, vamos ver como fazer operações.

Algumas pessoas costumam dizer que o R pode ser usado como uma calculadora, mas podemos fazer muito mais que isso. O primeiro passo é entender os tipos de operação disponíveis na linguagem.

Basicamente, podemos realizar qualquer operação com dados utilizando o R e armazená-las em uma ou mais variáveis. Podemos utilizar os seguintes tipos de operadores:

Operadores aritméticos:

    Soma (+), subtração (-), multiplicação (*), divisão (/), módulo (%%), exponenciação (^).

Operadores relacionais:

    Igual a (==), diferente de (!=), maior que (>), menor que (<), maior ou igual a (>=), menor ou igual a (<=).

Operadores lógicos:

    E (&&), OU (||), NOT (!).



### Operadores aritméticos

Os operadores aritméticos são:

* Soma (+)

* Subtração (-)

* Multiplicação (*)

* Divisão (/)

* Módulo (%%)

* Exponenciação (^)

Para fazer operações aritméticas, basta utilizar o símbolo correpondente diretamente com os números ou declarar variáveis. Abaixo seguem alguns exemplos.

*OBS*: Note que para imprimir os resultados precisamos utilizar o 'print()'. Em breve você conhecerá a fundo como esse tipo de comando funciona. Por enquanto, é só clicar em RUN.


In [ ]:
x <- 50
y <- 17

In [ ]:
print(x + y)

[1] 67


In [ ]:
print(x - y)

[1] 33


In [ ]:
print(x * y)

[1] 850


In [ ]:
print(x / y)

[1] 2.941176


In [ ]:
print(x %% y)

[1] 16


In [ ]:
print(x ^ y)

[1] 7.629395e+28


**Exercício**: Experimente trocar algumas vezes o valor de **x** e de **y** e rodar os comandos novamente.

*Aqui precisamos que você note que para trocar o valor da variável você necessita atribuí-la novamente clicando em RUN.*

### Operadores relacionais:

Os operadores relacionais fazem comparações entre as variáveis ou dados que estão carregados no sistema. São eles:

* Igual a (==)
* Diferente de (!=)
* Maior que (>)
* Menor que (<)
* Maior ou igual a (>=)
* Menor ou igual a (<=).

Para exemplificar vamos utilizar as mesmas variáveis carregadas acima. Note que diferente dos valores impressos nos exemplos anteriores, aqui o R vai apenas dizer se a comparação é verdadeira (TRUE) ou falsa (FALSE). É o que chamamos de operação booleana.

In [ ]:
print(x == y)

[1] FALSE


In [ ]:
print(x != y)

[1] TRUE


In [ ]:
print(x > y)

[1] TRUE


In [ ]:
print(x < y)

[1] FALSE


In [ ]:
print(x >= y)

[1] TRUE


In [ ]:
print(x <= y)

[1] FALSE


**Exercício**: Experimente atribuir o valor de **x** e de **y** antes das operações relacionais e rodar os comandos novamente.

*Aqui precisamos que você note que é possível declarar novamente o valor de uma variável ao longo do mesmo script.* Isso é importante porque o R lê as linhas de código na sequência em que estão escritas e isso pode mudar os resultados obtidos em scripts e operações mais complexas.

### Operadores lógicos

Assim como os operadores relacionais, os operadores lógicos retornam apenas valores booleanos. São eles:

* E (&)
* OU (|)
* NOT (!)

Se estamos trabalhando com conjuntos, o R também possui operadores específicos para fazermos testes lógicos.

*x pertence a y*
* x %in% y

*x ou y são verdadeiros (apenas um deles)*
* xor(x, y)

In [ ]:
print(x & y)

[1] TRUE


In [ ]:
print(x | y)

[1] TRUE


In [ ]:
print(!y)

[1] FALSE


In [ ]:
# Atribuindo um conjunto de dado a uma variável chamada 'z'
z <- c(12,3,5,9)

In [ ]:
x %in% z
y %in% z

[1] FALSE

[1] FALSE

## Tipos e estruturas de dados


O R possui uma variedade de tipos de dados que podem ser utilizados. Até agora vimos apenas os tipos numéricos, mas basicamente qualquer representação que conseguirmos fazer com um caracter é possível ser tratada e abordada. Abaixo segue uma lista com os principais tipos que utilizamos no dia-a-dia:

1. **Numéricos**:
   - Inteiros (`integer`): Números inteiros, como 1, 2, -3, 0.
   - Reais (`numeric` ou `double`): Números reais, como 3.14, -2.5, 0.0.

    Os números reais devem ser representados utilizando '.' e não ',' para marcar as casas decimais. Por exemplo, x <- 2.5 e não x <- 2,5.

    ```
    x <- 25,98372652 (errado)
    x <- 25.98372652 (correto)

    ```
    Para declarar números inteiros é necessário colocar um L logo depois do número.
    ```
    num_parti <- 25 (errado)
    num_parti <- 25L (correto)

    ```


2. **Caracteres** (`character`):
   - Strings de texto, como "Olá, mundo!", 'R é incrível'.

    Os caracteres sevem para que possamos trabalhar com textos. Eles devem sempre ser colocar entre aspas.

    ```
    x <- Olá, mundo! (errado)
    x <- 'Olá, mundo!' (correto)
    x <- '123' (Aqui o 123 será tratado como texto.)

    ```

3. **Lógicos** (`logical`):
   - Valores booleanos, como `TRUE` e `FALSE`.

    Como vimos na seção anterior, os tipos lógicos servem para fazermos comparações entre valores ou mesmo atribuir esse valor a um campo ou unidade de conjunto.

    ```
    exame_realizado <- TRUE
    exame_realizado <- FALSE

    ```

4. **Fatores** (`factor`):
   - Variáveis categóricas, como "masculino", "feminino", "alto", "baixo".

    Utilizamos fatores quando vamos trabalhar categorização de colunas em uma dataframe. Por exemplo, queremos atribuir um nome a determinada escala de valor.

    ```
    genero <- factor(c("Masculino", "Feminino", "Masculino"))

    ```

5. **Datas e Horas**:
   - `Date`: Datas, como "2023-04-18".
   - `POSIXct` e `POSIXlt`: Datas e horas, como "2023-04-18 14:30:00".

    As datas são variáveis importantes em qualquer estudo em que precisamos estabelecer um período de tempo. Por exemplo, quando queremos descobrir qual foi a precipitação ou a temperatura em determinado mês ou ano. Por isso precisam ser cuidadosamente tratadas antes que realizemos qualquer cálculo.
    O R possui três funções básicas (nativas) para tratamento de datas. `date` nos permite tratar apenas as datas sem hora ou registro de sistema (timestamp). POSIXct e POSIXlt nos permitem utilizar aqueles registros das bases que veem compostos com hora, dia e outras informações de sistema. Muitas vezes precisamos saber quando um determinado evento ocorreu ou foi registrado, em geral, esse registro vem com essas informações suplementares, também conhecidas como `timestamp`.
    Ao longo do curso você terá contato com vários exemplos que incluem datas e timestamps.

    ```
    data_nascimento <- as.Date("2022-03-15")

    reg_atendimento <- as.POSIXct("2023-10-27 15:30:00")

    ```
    
    A diferença entre ***POSIXct*** e ***POSIXlt*** está na forma como eles representam datas e horas no R.

    *POSIXct*:
        POSIXct é um tipo de dado numérico que armazena datas e horas como o número de segundos desde 1º de janeiro de 1970, 00:00:00 UTC (Tempo Universal Coordenado).
        Esse formato é amplamente utilizado em sistemas computacionais, pois é fácil de manipular e calcular diferenças entre datas.
        POSIXct é o formato recomendado para a maioria das aplicações, pois é mais eficiente em termos de memória e processamento.

    *POSIXlt*:
        POSIXlt é um tipo de dado de lista que armazena datas e horas de uma forma mais detalhada, com campos separados para ano, mês, dia, hora, minuto, segundo, etc.
        Esse formato é útil quando você precisa acessar ou manipular componentes individuais da data e hora, como extrair o dia da semana ou o mês.
        POSIXlt é mais flexível, mas também consome mais memória e é menos eficiente em termos de processamento do que POSIXct.

        Veja os exemplos de uso abaixo:

    ```
    # Criando uma data e hora usando POSIXct
    data_hora_ct <- as.POSIXct("2023-04-18 14:30:00")
    class(data_hora_ct)
    # [1] "POSIXct" "POSIXt"
    # Criando uma data e hora usando POSIXlt
    data_hora_lt <- as.POSIXlt("2023-04-18 14:30:00")
    class(data_hora_lt)
    # [1] "POSIXlt" "POSIXt"

    # Acessando componentes individuais
    data_hora_lt$year
    # [1] 123
    data_hora_lt$mon
    # [1] 3
    data_hora_lt$mday
    # [1] 18
    ```

    Note que aqui você terá possibilidades diferentes com os dados. No POSIXct é possível fazer a localização do dado por meio do registro no sistema, enquanto no POSIXlt você poderá extrair de forma eficiente dia, mês e ano sempre precisar criar colunas extra em seu conjunto de dados.


6. **Vetores** (`vector`):
   - São coleções de valores do mesmo tipo de dado. Utilizamos a seguinte marcação (função) para declará-los.
   
   ```   
    numeros <- c(1, 2, 3, 4, 5)
    letras <- c('a', 'b', 'c')

   ```

**ESTRUTURAS DE DADOS**

7. **Matrizes** (`matrix`):
   - São estruturas bidimensionais de dados do mesmo tipo. Se pensarmos em outros programas de computador, podemos dizer que são como tabelas, possuem um número específico de linhas e colunas.
   
   Observe o exemplo abaixo:

   ```   
    matriz <- matrix(c(1, 2, 3, 4, 5, 6), nrow = 2, ncol = 3)

   ```

  Ao executar esse script, o R criará uma variável chamada matriz que contém os números de 1 a 6 armazenados em uma tabela com 2 linhas (nrow) e três colunas (ncol). Ficando da seguinte forma:

  ```
       [,1] [,2] [,3]
   [1,] 1    3    5
   [2,] 2    4    6

  ```
  *OBS:* Até aqui, você só precisa compreender que essa é função disponível no R, ao longo do curso e com o uso da ferramenta vamos aprofundando nos usos das matrizes e outras estruturas de dados.  


8. **Listas** (`list`):
   - São coleções de objetos de diferentes tipos de dados. Enquanto as matrizes só atribuição dos mesmos tipos de dados, as listas nos permitem realizar a mesma tarefa "mesclando" tipos diferentes de dados. Observe o exemplo abaixo:

    ```   
    programas_saude <- list("Programa de Vacinação" = list(objetivo = "Aumentar a cobertura vacinal", publico_alvo = "Crianças e idosos", numero_de_participantes = 40L, orcamento = 35.5), "Programa de Prevenção ao Tabagismo" = list(objetivo = "Reduzir o número de fumantes", publico_alvo = "Adolescentes e adultos", numero_de_participantes = 60L, orcamento = 50.75), "Programa de Saúde Mental" = list(objetivo = "Melhorar o acesso a serviços de saúde mental", publico_alvo = "População em geral" numero_de_participantes = 50L, orcamento = 25.5))
    
    ```   
    Note que agora é possível colocar diferentes categorias atribuídas a uma mesma variável, incluindo qualquer tipo de dado. Em outras linguagens de programação, como o Python, isso também é chamado de dicionário de dados. Com eles podemos interagir com diversas categorias, nomear elementos dentro de um dataset e outras funcionalidade muito úteis para preparação e manipulação de dados.
    
    Se você imprimir a variável `programas_saude` imprimir no console utilizando print() verá a seguinte saída.
    
    ```
    $`Programa de Vacinação`
    $`Programa de Vacinação`$objetivo
    [1] "Aumentar a cobertura vacinal"

    $`Programa de Vacinação`$publico_alvo
    [1] "Crianças e idosos"

    $`Programa de Prevenção ao Tabagismo`
    $`Programa de Prevenção ao Tabagismo`$objetivo
    [1] "Reduzir o número de fumantes"

    $`Programa de Prevenção ao Tabagismo`$publico_alvo
    [1] "Adolescentes e adultos"

    $`Programa de Saúde Mental`
    $`Programa de Saúde Mental`$objetivo
    [1] "Melhorar o acesso a serviços de saúde mental"

    $`Programa de Saúde Mental`$publico_alvo
    [1] "População em geral"
    ```


9. **Dataframes** (`data.frame`):
   - Enquanto as listas e matrizes se assemelham às tabelas, os dataframes são semelhantes às planilhas. A diferença entre uma lista e um dataframe é de que cada categoria vai corresponder a uma coluna e não a uma linha de dados. Dessa maneira, é possível armazenar uma quantidade maior de dados, homegeneizá-los e transformá-los. Observe o exemplo abaixo:
    ```   
    pacientes <- data.frame(
      nome = c("João Silva", "Maria Oliveira", "Pedro Souza", "Ana Pereira"),
      idade = c(45, 32, 58, 27),
      diagnostico = c("Diabetes", "Hipertensão", "Asma", "Gripe"),
      tratamento = c("Medicação", "Medicação", "Medicação", "Repouso"),
      data_admissao = as.Date(c("2023-04-01", "2023-03-15", "2023-02-20", "2023-04-05")))
    
    ```   
    Note que agora todos os nomes estão agrupados na categoria nome, assim como as idades, diagnósticos e assim por diante. Quando declaramos uma lista fazermos o caminho contrário, todas as características do objeto que estamos descrevendo ficam agrupadas junto a ele.

    Se você imprimir a variável `pacientes` imprimir no console utilizando print() verá a seguinte saída.

    ```
                nome idade diagnostico tratamento data_admissao
    1     João Silva    45    Diabetes  Medicação    2023-04-01
    2 Maria Oliveira    32 Hipertensão  Medicação    2023-03-15
    3    Pedro Souza    58        Asma  Medicação    2023-02-20
    4    Ana Pereira    27       Gripe    Repouso    2023-04-05
    ```
    Agora é possível ver os dados tabulados e organizados em formato de linhas e colunas.

**EXPRESSÕES E FUNÇÕES**

10. **Expressões** (`expression`):
    - O R nos permite escrever a representação de expressões matemáticas de forma nativa. Basta usar a função expression.
11. **Fórmulas** (`formula`):
    - Representação de modelos estatísticos também é realizada com facilidade, por meio da expressão formula.
12. **Funções** (`function`):
    - Enquanto linguagem de programação é possível criar conjuntos de scripts que podem ser chamados em conjunto apenas com um comando. Sabe o print(), o c(), o list(), o vector() e o data.frame() são todos funções que já vem instaladas com o R. A parte mais interessante é que você pode criar as suas próprias funções, de acordo com as suas necessidades. Mais adiante vamos nos aprofundar nisso.

**Exercício**:

1. Imagine que você está realizando a gestão de três programas de saúde em uma secretaria estadual. Utilizando o RStudio, escreva um script que realize as seguintes tarefas:

   Dados de entrada (variáveis):
   - Nome dos programas
   - Objetivos dos programas
   - Público-alvo dos programas
   - Orçamento anual de cada programa
   - Datas de início e fim de cada programa

   Tarefas:
   - Criar uma lista com as informações de cada programa
   - Calcular o orçamento total destinado a todos os programas
   - Identificar o programa com o maior orçamento
   - Listar os programas que estão em andamento atualmente

  Desafio:
   - Exibir o programa com a data de início mais recente

  *Orientações:*
    
    - O primeiro objetivo desta atividade é treinar a atribuição de tipos de dados às variáveis e contantes. Então, declare uma variável separada para cada dado de entrada solicitada, utilizando um tipo de dados adequado ao que deseja atribuir ou representar. Exemplo, se a variável é orçamento o tipo de dados deverá ser numérico.
    - O segundo objetivo é exercitar o uso de operadores. Então, busque testar os operadores que melhor se adequam às tarefas de processamento solicitadas. Por exemplo, a quantidade total de orçamento normalmente é a soma do valor de todos os programas.
    - Procure fazer os exercícios antes de procurar por soluções prontas. Lembre-se que não estamos realizando uma "prova" mas um treinamento. Aqui o processo de aprendizado é mais importante que o resultado pronto.


# Algumas dicas úteis

## Lista de Constantes Nativas do R

Como vimos acima as contantes são valores que não se modificam ao longo do processamento. Facilitam o processamento e a escrita dos códigos. Apenas esse tema já daria por si só um capítulo ou mesmo um curso.

Abaixo deixamos uma lista de contantes que já vem junto com o R que podem facilitar muito o nosso trabalho. Aos poucos, vá pesquisando como utilizá-las e experimentando compor os seus scripts com elas. Dois exemplos úteis, são as contantes de valor ausente (NA) e valos não numérico (NaN), assim como as contantes que fazem a leitura de caracteres latinos (encoding).


1. **Constantes matemáticas**:
   - `pi`: Valor de pi (aproximadamente 3.14159)
   - `e`: Valor da constante de Euler (aproximadamente 2.71828)

2. **Constantes lógicas**:
   - `TRUE`: Valor lógico verdadeiro
   - `FALSE`: Valor lógico falso
   - `NA`: Valor ausente (Not Available)
   - `NaN`: Valor não numérico (Not a Number)

3. **Constantes de precisão**:
   - `.Machine$double.eps`: Menor número positivo `x` tal que `1 + x != 1`
   - `.Machine$double.xmin`: Menor número positivo normalizado
   - `.Machine$double.xmax`: Maior número finito positivo

4. **Constantes de sistema**:
   - `.Platform$OS.type`: Tipo de sistema operacional (Windows, Unix, etc.)
   - `.Platform$GUI`: Interface gráfica do usuário (X11, Aqua, RStudio, etc.)
   - `.Platform$endian`: Ordem de bytes do sistema (little-endian ou big-endian)

5. **Constantes de datas e horas**:
   - `Sys.Date()`: Data atual do sistema
   - `Sys.time()`: Hora atual do sistema
   - `as.Date("1970-01-01")`: Data de referência (1 de janeiro de 1970)

6. **Constantes de codificação de caracteres**:
   - `Encoding("UTF-8")`: Codificação de caracteres UTF-8
   - `Encoding("Latin-1")`: Codificação de caracteres Latin-1 (ISO-8859-1)


*OBS:* Esses são apenas alguns exemplos das contantes mais utilizadas.

# Materiais e referências



- [Livro MQ](https://bookdown.org/eugeniadoria/CursoMQ/)

- [Documentação do R](https://cran.r-project.org/doc/manuals/r-release/R-intro.html)